### DISCUSSION TITLE
Docker: How to keep configuration between container rebuilds

### DISCUSSION THREAD

**Discussion by jacgl** [jcorporation/myMPD#659] [Q&A] [answered]
Hi, 
I am just exploring version 9.0.3 and playing with docker image, overall very nice piece of software. 
Now, trying some ideas of mapping volumes to keep the config...
And this is not a straight way. The issue is with mympd user, and probably access rights for him. The config (/var/lib/mympd/config) is only for root, other files are changed for mympd:mympd. Even docker volume created by mympd container itself and mounted to /var/lib/mympd does not work as it should, because of rights. Any hints?
uid and gid are already used in debian

**Comment by jcorporation** [comment]
https://jcorporation.github.io/myMPD/installation/docker

Usage of  puid and pgid should solve your problem.

**Reply by jacgl** [reply]
This is what I tried at the very beginning
      - PUID=1000
      - PGID=1000
but unfortunately the setting does not work, there is still following line in the log
NOTICE   mympd     Droping privileges to user "mympd"

The result depends on rights given to dir mapped by docker-compose. If it is 755 root:root only config dir is created, and mympd tries to create ssl certificates even for MYMPD_SSL=false. 
NOTICE   mympd     Starting myMPD 9.0.3
NOTICE   mympd     Listening on http://0.0.0.0:80
NOTICE   mympd     Listening on https://0.0.0.0:443
NOTICE   mympd     Serving files from "/var/lib/mympd/empty"
NOTICE   mympd     Droping privileges to user "mympd"
ERROR    mympd     SSL cert dir: creating "/var/lib/mympd/ssl" failed
ERROR    mympd     Permission denied
NOTICE   mympd     Work dir: "/var/lib/mympd"

If the rights are 777, the whole set is created, but config is default. The env variables defined for docker-compose do not have any impact.

**Comment by jcorporation** [comment]
> NOTICE mympd Droping privileges to user "mympd"

This is expected: https://docs.linuxserver.io/general/understanding-puid-and-pgid

Environment variables are ignored after first startup: https://jcorporation.github.io/myMPD/configuration/

All subfolders in /var/lib/mympd must be writeable by the mympd user (except config).

I am not a docker expert, but other users run myMPD successfully with docker.

**Reply by jacgl** [reply]
In general, myMPD runs in container perfectly, anyway keeping the config is a pain, Ever worse, that myMPD modifies not only the owner, but the rights to files and dirs in /var/lib/mympd/
It would be much better if PUID/GUID was never ignored.

**Comment by jcorporation** [comment]
myMPD does not use puid/guid that’s docker.

I am not sure if I understand your problem. Is the problem that uids are changing between builds or what is the cause of the permission problems?

**Reply by jacgl** [reply]
Not really... The issue is that newly created myMPD container (1st run) is not able to create config files (i.e. config, empty, pics, scripts, smartpls, ssl, state dirs and proper set of files inside) form env settings, when empty host dir is mapped to /var/lib/mympd. Instead it creates default config files. The same when docker volume is mapped to /var/lib/mympd
When there is no mapping, all the files are created correctly according to env settings, but inside container. 
Seems like I found workaround. Let myMPD create defaults, then stop container, edit the files (for example http_port, ssl to false), and start container. When container is removed and recreated, the settings are kept as well.
Not tried anything before 9.0.3. Currently mympd user is always 100:101, and the files are created with this ownership (which looks strange in debian, because user 100 and group 101 already exist as built-in)

**Comment by jcorporation** [comment, marked as answer]
Now I understand your problem. I looked at the initialization code and found an issue. With commit 3951b9bfddac77b215e423492a533b3055f25c28 myMPD chowns also existing workdirs. This should fix the initial startup problem with mounted /var/lib/mympd volume. 

Edit: v9.0.4 fixes this issue
